keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -345,24 +345,31 @@ kind: Service
 metadata:
   name: my-service
 spec:
+  type: ClusterIP
+  ClusterIP: 1.2.3.4
   ipFamily: IPv6
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
+status:
+  preferredIPFamily:
+    - IPv46
 ```
 
 Dual-stack services (having both an IPv4 and IPv6 cluster IP) are currently out of scope of this proposal. Should there be use-cases where dual-stack Services are needed then we can revisit.
 
 #### Type Headless services
-
-\<TBD\>
+Headless services will operate using the following conditions:
+1. api-server does not default `ipFamily` field. User may optionally to choose to set it.
+2. If `ipFamily` field is set by the user, then the value of `preferredIPFamily` will follow accordingly. Example: setting `ipFamily` to `IPv6` then `preferredIPFamily` will have `IPv6` as a value.
+3. If `ipFamily` field is not set, then the value of `preferredIPFamily` will carry the values set in `-- service-cluster-ip-range`, which can be a single value in even in a dual stack cluster.
```

Comment:
nit!
```suggestion
3. If `ipFamily` field is not set, then the value of `preferredIPFamily` will carry the values set in `--service-cluster-ip-range`, which can be a single value in even in a dual-stack cluster.
```

Reply:
Wait, you're allowed to have single-stack `--service-cluster-ip-range` in a dual-stack cluster? I didn't realize that. (That ought to be called out explicitly if so since network plugins have to deal with that.)

Reply:
I will make sure it is clear in the KEP and our docs, do you recommend other places where we should make that call out?

Reply:
well, like, I specifically added a unit test to ensure we reject this as an invalid configuration in ovn-kubernetes: https://github.com/ovn-org/ovn-kubernetes/pull/1189/files#diff-173cba3f6f9206af564ac9ed53992a91R982 :slightly_smiling_face:

Reply:
@smarterclayton was the one that clarified this https://github.com/kubernetes/kubernetes/pull/86895#issuecomment-571725086 ,

> >is it possible to enable dual-stack without configuring two service CIDRs? I mean configuring only a service-cidr for IPv6 and not for IPv4?
> 
> The feature gate flag allows dual stack config. Enabling the flag doesn’t require you to use dual stack.

Reply:
The text says "`--service-cluster-ip-range` ... can be a single value even in a dual-stack cluster". It should not say "a dual-stack cluster" if it means "any cluster in which the `IPv6DualStack` feature gate is active, regardless of configuration".

Saying "you can enable the `IPv6DualStack` feature gate and not be obligated to configure dual-stack networking" is unnecessary, because that's a hard requirement of the feature process. (Otherwise we'd never be able to get rid of the feature gate and declare the feature GA.)

So if the text means "`--service-cluster-ip-range` can be a single value even when `IPv6DualStack` is enabled" then that doesn't need to be said. But if it means "`--service-cluster-ip-range` can be single-stack even when the pod network is dual-stack", then that's surprising.

Reply:
thanks for the clarification, I was always assuming 
> --service-cluster-ip-range can be a single value even when IPv6DualStack is enabled

Reply:
@aojea  Your assumption is correct.